### ISSUE TITLE
When I publish a data to pubsub topic, I have a memory leak on kubernetes pod.

### ISSUE THREAD

**Issue by fodrh1201** [googleapis/python-pubsub#395]
Each time a public request is made, the memory in the pod of Kubernetes increases. Removing the publishing part does not increase memory. 
The pod is continuously restarted due to OOM.

#### Environment details

  - OS type and version: Debian GNU/Linux 10 
  - Python version: 3.8.7
  - pip version: 21.0.1
  - `google-cloud-pubsub` version: 2.4.1

#### Code example

```python

class PublishToPubsubCallback(Callback):
    def __init__(self, pod_name, project_id, topic_id, preprocess=[]):
        super(PublishToPubsubCallback, self).__init__(pod_name, preprocess=preprocess)
        self.project_id = project_id
        self.topic_id = topic_id
        self.publisher = pubsub_v1.PublisherClient()
        self.topic_path = self.publisher.topic_path(project_id, topic_id)
        
    def __call__(self, data):
        pub_data = json.dumps(data).encode('utf-8')
        self.publisher.publish(self.topic_path, pub_data)

callback = PublishToPubsubCallback(pod_name, project_id, topic_id)

async def run():
    while True:
        data = await watch_websocket_data()
        callback(data)

loop = asyncio.get_event_loop()
loop.run_until_complete(run())

```

![스크린샷 2021-04-23 오후 4 58 16](https://user-images.githubusercontent.com/6869272/115839538-b6bdce80-a455-11eb-97a0-ebf97b1513fb.png)

**Comment by plamut** [comment]
@fodrh1201 Thanks for the report. Would it be possible to provide a minimal runnable self-contained code sample? We are not aware of any publisher memory leaks, thus we would first like to exclude any code outside of the client as a possible reason.

BTW, one thing that caught my eye is the mutable default argument to `PublishToPubsubCallback`'s `__init__()` method, i.e. `preprocess=[]`. That empty list is instantiated only once at method definition time, and is shared among all instances of `PublishToPubsubCallback`. If the instances append anything to that list, it will continuously grow, consuming more and more memory.

In the code sample the callback is actually instantiated only once, but might nevertheless be worth checking out if that's also the case in the actual code.

**Comment by fodrh1201** [comment]
This is python code.

```python
import json
from google.cloud import pubsub_v1
import asyncio
import logging
from datetime import datetime


class Callback:
    
    def __init__(self, pod_name, is_work_at_stop=False, preprocess=[]):
        self.is_work_at_stop = is_work_at_stop
        self.pod_name = pod_name
        self.preprocess = preprocess
    
    def __call__(self, data):
        for preprocess in self.preprocess:
            data = preprocess(data)
        return data


class PublishToPubsubCallback(Callback):
    
    def __init__(self, pod_name, project_id, topic_id, preprocess=[]):
        super(PublishToPubsubCallback, self).__init__(pod_name, preprocess=preprocess)
        self.project_id = project_id
        self.topic_id = topic_id
        self.publisher = pubsub_v1.PublisherClient()
        self.topic_path = self.publisher.topic_path(project_id, topic_id)
        
    # @profile
    def __call__(self, data):
        data = super().__call__(data)
        pub_data = json.dumps(data).encode('utf-8')
        self.publisher.publish(self.topic_path, pub_data,
                            origin=self.pod_name, data_type=data['data_type'],
                            symbol=data['symbol'], client_id=data['client_id'],
                            timestamp=str(data['timestamp']),
                            market_type=data['market_type'])

data = {
    'data_type': 'orderbook',
    'client_id': 'binance',
    'symbol': 'BTC/USDT',
    'timestamp': '1619494646034',
    'market_type': 'spot',
    'asks': [[53482.36, 2e-06], [53493.34, 0.025379], [53493.64, 0.01412]],
    'bids': [[53482.35, 2.882399], [53479.18, 0.000373], [53477.2, 0.406]]
}

callback = PublishToPubsubCallback('test', 'bx-datafeed', 'test')
async def run():
    while True:
        callback(data)
        await asyncio.sleep(0.1)

loop = asyncio.get_event_loop()
loop.run_until_complete(run())
```

this is the result of above code.

![스크린샷 2021-04-28 오후 1 18 14](https://user-images.githubusercontent.com/6869272/116346012-4384d580-a824-11eb-956e-9a83703c3d2f.png)



my docker file is below.
```
FROM python:3.8
WORKDIR /app
ADD . .
RUN pip install argparse numpy google-cloud-pubsub==2.4.1 six==1.15.0

ENV PYTHONBUFFERED 1
```

**Comment by plamut** [comment]
I ran a slightly modified version of the example locally and measured memory consumption. I put the main code into the `main()` function (which was profiled), and changed the loop to run for 30_000 iterations. This is the graph:

![publisher_mem_usage](https://user-images.githubusercontent.com/1235663/116519091-586b7280-a8d1-11eb-9909-582638a6831c.png)

It appears that the memory is indeed slowly but surely leaking. Can't yet say why, will investigate further.

A possible workaround is to modify the code to run in a subprocess (for a bounded number of iterations), and creating a new subprocess when the previous one finishes. Not ideal, of course, but at least the memory will be released at regular intervals.

**Comment by plamut** [comment]
It turns out the issue is reproducible even with a straightforward example that repeatedly publishes a trivial message in a loop.

```py
import time

from google.cloud import pubsub_v1


PROJECT_ID = "set-project-id"
TOPIC_ID = "set-topic-id"


def main():
    publisher = pubsub_v1.PublisherClient()

    topic_path = publisher.topic_path(PROJECT_ID, TOPIC_ID)
    print("Starting the main loop...")

    for i in range(15_000):
        if not i % 1000:
            print(f"{i: 5}: publishing a message")

        publisher.publish(topic_path, b"x")
        time.sleep(0.01)

    print("Main loop complete!")


if __name__ == "__main__":
    main()
```

Using `tracemalloc` and taking a snapshot every 1000 iterations showed that  by far the most memory was allocated in the `threading` module, but the (Python) traceback was limited:
```
130.9921875 new KiB, 1991.53125 total KiB, 3724 new, 56645 total memory blocks: 
  File "/usr/local/lib/python3.8/threading.py", line 890
    self._bootstrap_inner()
  File "/usr/local/lib/python3.8/threading.py", line 918
    self._set_tstate_lock()
  File "/usr/local/lib/python3.8/threading.py", line 908
    self._tstate_lock = _set_sentinel()
```
It appears that the these allocations are initiated by something outside of Python , maybe gRPC?

@lidizheng Do you know if there are currently any confirmed or suspected leaks in gRPC? Do you have any free cycles to take a look in the near future? Is there anything else I can provide on top of the above minimal example?

**Versions:**
```
Python 3.8
Ubuntu 18.04.4 LTS

google-cloud-pubsub==2.4.1 
grpcio==1.36.1
google-api-core==1.26.0
google-auth==1.28.0
```

**Comment by lidizheng** [comment]
@plamut  I will take a look today or tomorrow. Can you also provide the numbers you see? Like for each run, how many more memory consumed, what's the eventual memory consumption, and what's the virtual memory usage?

**Comment by plamut** [comment]
@lidizheng This is the `memoryprofiler` run with the `@profile` decorator on the `main()` function:

![mem_usage](https://user-images.githubusercontent.com/1235663/116614680-8769fe80-a93a-11eb-9be2-effa0ba9b487.png)

The memory usage initially jumps to almost 70 MiB and the ngradually increases to almost 79 MiB. I'm attaching the raw profiling data below (renamed the `.dat` suffix to `.txt`):

[mprofile_data.txt](https://github.com/googleapis/python-pubsub/files/6401499/mprofile_data.txt)

**Comment by plamut** [comment]
Found a possible fix! If the client [commit thread](https://github.com/googleapis/python-pubsub/blob/4fe72cad610362a3147ce71ecadf5eaaf6346981/google/cloud/pubsub_v1/publisher/client.py#L378-L380) is marked as a daemon, the leak disappears. Can't say why, though.

**Comment by plamut** [comment]
For additional context, I found this [CPython issue](https://bugs.python.org/issue39074) and the `tracemalloc` traceback from the reproduction looks _very_ familiar:
```
New Bytes: 36000	Total Bytes 36000	New blocks: 1000	Total blocks: 1000: 
  File "/usr/local/lib/python3.7/threading.py", line 890
    self._bootstrap_inner()
  File "/usr/local/lib/python3.7/threading.py", line 914
    self._set_tstate_lock()
  File "/usr/local/lib/python3.7/threading.py", line 904
    self._tstate_lock = _set_sentinel()
```

The issue has been closed as a duplicate of [another issue](https://bugs.python.org/issue37788), which is currently still open, and reportedly affects Python 3.8+.

**Comment by fodrh1201** [comment]
@plamut Thank you!! I solved the problem thanks to you.
